### ISSUE TITLE
Threads do not die

### ISSUE THREAD

**Issue by cmungall** [geneontology/minerva#28]
The latency with Minerva seems to grow and grow becoming increasingly problematic.

Option 1, switch off reasoning

Option 2 from @kltm

> The other option is just to have something like a nightly cycle for the
> time being.
> 
> What I think is going on, so far:
> 
> swdev@toaster:~$ ps -AlF | grep minerva-cli.jar | wc
>      3      97    1352
> 
> Looking, this is the grep command, the shell script, and the actual java
> call. Now then, look at this:
> 
> swdev@toaster:~$ ps -AlFT | grep minerva-cli.jar | wc
>    765   29818  479144
> 
> That's 765...what is this? They are all child...threads? I'm not
> entirely sure what is going on here, but the number looks /something/
> like every command to the server is generating a new one of whatever
> these are. Any thoughts? Reasoner issues? Something odd with how the
> calls are being taken by the web server?
> 
> We can start experimenting about when exactly these are generated, but
> it would seem that there is some kind of leak at the thread level.

**Comment by cmungall** [comment]
One possibility is that HermiT is not tidying up after itself. We could experiment with some different slme wrapped reasoners here

**Comment by cmungall** [comment]
Almost unusable now, need to restart

**Comment by kltm** [comment]
TL;DR: Likely in the reasoner/model neck of the woods, not with the webapp. 

Starting:
```
~/local/src/git/noctua$:) ps -eLF | grep minerva-cli.jar | wc -l
24
```
Refresh to view model (get, model):
```
~/local/src/git/noctua$:) ps -eLF | grep minerva-cli.jar | wc -l
29
```
Repeats (get, model) do not increase the count.

Go back to landing page (get, meta); an increase of 1, up to 30, but after a breif pause, back down to 29; repeats do nothing.

Back to the original model (Scratch, btw), using ubernoodle added an individual. No noticeable change.

Added an edge between the new individual and something else: drops to 25, then back up to 29.

Go to a different model, this takes us up to 33...so 4 threads per model? At (/ (- 765 25) 4) => 185, being that we don't have 185 models (last time I checked), that's not it.

Opening a new browser, as a new user, to the same model, does nothing.

Saving causes the drop and return (29 to 33).

Added another edge and individual, no change.

Killed minerva, down to 1--the grep command again.

So there must be /some/ command to minerva that is getting us up to that high number. I'm assuming that the high number is causing the slowdown (which would be nice to double check).

**Comment by jnguyenx** [comment]
I suggest to run VisualVM on minerva to have more insights of what's going on. Try to sample the method calls, on the memory tab if I recall correctly.

**Comment by cmungall** [comment]
Also, if this is not something related to running in jetty, then it should be easy to replicate this with a junit stress/integration test that performs multiple operations. Based on @kltm's analysis above it might be useful to test cycling through models.

I'm still not clear about the role of the reasoner in this. Yesterday I thought we restarted minerva with reasoning off, yet I see inferences there...

Is there a minerva health check and/or config query call?

**Comment by kltm** [comment]
May not be directly related to performance: production is up to 12,600 threads.

**Comment by kltm** [comment]
No jetty should be involved here--isn't jersey-server just it's own thing?
We did not turn of the reasoner; I think we just turned off one of the settings for it.

**Comment by kltm** [comment]
Found the reasoner flags in minerva-server/src/main/java/org/geneontology/minerva/server/StartUpTool.java, restarting with reasoner off; will report back later.

**Comment by kltm** [comment]
Well, after a few hours, we're at 43 still. Will watch some more, but this seems like a huge improvement so far.

**Comment by kltm** [comment]
And by the way, when minerva was killed at ~13k threads, it has also managed to finally eat through all of its allocated 32GB memory.

**Comment by kltm** [comment]
Over 24hrs use, at 116, and this is also after a very large number of manual repl operations across a whole bunch of models.

**Comment by DoctorBud** [comment]
I've been looking into this and have a reproducible way to poke minerva-server and get it to generate 8 new Elk reasoner threads, without throwing away the first 8. So you get 8 new threads every time. Either a Factory isn't being reused, or it's not doing it's job. I'm investigating.

Here's the URL, fyi. I haven't decomposed it to make it more readable:

```
curl -v 'http://localhost:6800/m3BatchPrivileged?intention=action&requests=%5B%7B%22entity%22%3A%22model%22%2C%22operation%22%3A%22add%22%2C%22arguments%22%3A%7B%7D%7D%5D&uid=http%3A%2F%2Forcid.org%2F0000'
```

I also discovered that you can send a signal to a Java process to get it to log its threads, which is sometimes handier than using `jvisualvm`:

```
> ps |grep minerva-server.jar
 4803 ttys000    6:27.67 /usr/bin/java -Xmx2G -jar ./minerva-server/bin/minerva-server.jar -g ./cache/merged.owl -f ../noctua-models/models --port 6800 -c ./cache/catalog-v001.xml --use-golr-url-logging --slme-elk --use-golr-url-logging --use-golr-url-logging exit
> kill -3 4803
```

**Comment by DoctorBud** [comment]
(copied from gitter...)

I think I found at least one bug that is responsible for the thread explosion. Basically, a missing `reasoner.dispose()` in `JsonOrJsonpBatchHandler.java`. I added a `dispose` at the end of the workhorse  `m3batch` method like this:

```
		if (useReasoner) {
			if (reasoner != null) {
				reasoner.dispose();
			}
		}

		return response;
	}

	public static void initResponseData(JsonModel jsonModel, ResponseData data) {
```

and the Elk-Reasoner threads were disposed of by the end of the run.

Looking at `JsonOrJsonpSeedHandler.java`, I see that the `dispose()` is already present, AND it is in a `finally` clause. I’m building a PR with my provisional fix. Going to see about making a unit test to detect thread explosion, but  I think that is really a different type of unit test.

**Comment by DoctorBud** [comment]
I have a Pull Request here:

https://github.com/geneontology/minerva/pull/30

that adds the dispose() call.

**Comment by hdietze** [comment]
The proposed change is a stop-gap until we can fix the underlying issues. Even with disposing of the reasoners, we will run into situations with too many threads (i.e. more than model creates a reasoners) and thrashing.

**Comment by hdietze** [comment]
Stepping through the ELK code, I can confirm that each ELK reasoner instance creates it's own thread pool. There is no way to override that behavior, see org.semanticweb.elk.reasoner.ReasonerFactory#createReasoner

**Comment by balhoff** [comment]
@kltm @cmungall would you say this is no longer a problem? I'm not sure anything changed with ELK, but we aren't actively using it now.

**Comment by kltm** [comment]
Those were exciting times, but alas over.
